src/NHibernate/Impl/SqlQueryImpl.cs

```diff
@@ -296,7 +299,75 @@ protected internal override void VerifyParameters()
 					}
 				}
 			}
+		}
+
+		/// <inheritdoc />
+		protected internal override void VerifyParameters(bool reserveFirstParameter)
+		{
+			ComputeFlattenedParameters();
+			base.VerifyParameters(reserveFirstParameter);
+		}
+
+		// Flattening parameters is required for custom SQL loaders when entities have composite ids.
+		// See NH-3079 (#1117)
+		private void ComputeFlattenedParameters()
+		{
+			var flattenedParameters = FlattenTypesAndValue(base.Types, base.Values.Cast<object>());
+			_flattenedTypes = flattenedParameters.Item1.ToList();
+			_flattenedValues = flattenedParameters.Item2.ToList();
+
+			Tuple<IEnumerable<IType>, IEnumerable<object>> FlattenTypesAndValue(ICollection<IType> types, IEnumerable<object> values)
+			{
+				var resultTypes = new List<IType>(types.Count * 2);
+				var resultValues = new List<object>(types.Count * 2);
+				using (var valuesEnumerator = values.GetEnumerator())
+				{
+					foreach (var type in types)
+					{
+						var actualType = type;
+						valuesEnumerator.MoveNext();
+						var value = valuesEnumerator.Current;
+						if (type is EntityType entityType)
+						{
+							actualType = entityType.GetIdentifierType(session);
+							value = entityType.GetIdentifier(value, session);
+						}
+
+						if (actualType is IAbstractComponentType componentType)
+						{
+							var flattened = FlattenTypesAndValue(
+								componentType.Subtypes,
+								componentType.GetPropertyValues(value, session));
+							resultTypes.AddRange(flattened.Item1);
+							resultValues.AddRange(flattened.Item2);
+						}
+						else
+						{
+							resultTypes.Add(actualType);
+							resultValues.Add(value);
+						}
+					}
+				}
+
+				return new Tuple<IEnumerable<IType>, IEnumerable<object>>(resultTypes, resultValues);
+			}
+		}
+
+		protected override IList Values => _flattenedValues ??
+			throw new InvalidOperationException("Flattened parameters have not been computed");
 
+		protected override IList<IType> Types => _flattenedTypes ??
+			throw new InvalidOperationException("Flattened parameters have not been computed");
+
+		public override object[] ValueArray()
+		{
+			return _flattenedValues?.ToArray() ??
```

Comment:
It could also be `Values.Cast<object>().ToArray()`. In which case it could be moved in the base class instead.

Reply:
Added a TODO for 6.0 for changing `Values` type to `IList<object>`.